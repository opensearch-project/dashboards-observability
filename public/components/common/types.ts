/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export interface FieldCapAttributes {
  type: string;
  searchable: boolean;
  aggregatable: boolean;
}

export interface FieldCapResponse {
  indices: string[];
  fields: Record<string, Record<string, FieldCapAttributes>>;
}
