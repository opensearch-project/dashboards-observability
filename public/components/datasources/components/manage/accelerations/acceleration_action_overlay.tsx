/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState } from 'react';
import { EuiOverlayMask, EuiConfirmModal, EuiCompressedFormRow, EuiCompressedFieldText } from '@elastic/eui';
import { CachedAcceleration } from '../../../../../../common/types/data_connections';
import {
  ACC_DELETE_MSG,
  ACC_VACUUM_MSG,
  ACC_SYNC_MSG,
  AccelerationActionType,
  getAccelerationName,
  getAccelerationFullPath,
} from './utils/acceleration_utils';

export interface AccelerationActionOverlayProps {
  isVisible: boolean;
  actionType: AccelerationActionType;
  acceleration: CachedAcceleration | null;
  dataSourceName: string;
  onCancel: () => void;
  onConfirm: () => void;
}

export const AccelerationActionOverlay: React.FC<AccelerationActionOverlayProps> = ({
  isVisible,
  actionType,
  acceleration,
  dataSourceName,
  onCancel,
  onConfirm,
}) => {
  const [confirmationInput, setConfirmationInput] = useState('');

  if (!isVisible || !acceleration) {
    return null;
  }

  const displayIndexName = getAccelerationName(acceleration);
  const displayFullPath = getAccelerationFullPath(acceleration, dataSourceName);

  let title = '';
  let description = '';
  let confirmButtonText = 'Confirm';
  let confirmEnabled = true;

  switch (actionType) {
    case 'vacuum':
      title = `Vacuum acceleration ${displayIndexName} on ${displayFullPath} ?`;
      description = ACC_VACUUM_MSG;
      confirmButtonText = 'Vacuum';
      confirmEnabled = confirmationInput === displayIndexName;
      break;
    case 'delete':
      title = `Delete acceleration ${displayIndexName} on ${displayFullPath} ?`;
      description = ACC_DELETE_MSG(displayIndexName);
      confirmButtonText = 'Delete';
      break;
    case 'sync':
      title = 'Manual sync data?';
      description = ACC_SYNC_MSG;
      confirmButtonText = 'Sync';
      break;
  }

  return (
    <EuiOverlayMask>
      <EuiConfirmModal
        title={title}
        onCancel={onCancel}
        onConfirm={() => onConfirm()}
        cancelButtonText="Cancel"
        confirmButtonText={confirmButtonText}
        buttonColor="danger"
        defaultFocusedButton="confirm"
        confirmButtonDisabled={!confirmEnabled}
      >
        <p>{description}</p>
        {actionType === 'vacuum' && (
          <EuiCompressedFormRow label={`To confirm, type ${displayIndexName}`}>
            <EuiCompressedFieldText
              name="confirmationInput"
              value={confirmationInput}
              onChange={(e) => setConfirmationInput(e.target.value)}
            />
          </EuiCompressedFormRow>
        )}
      </EuiConfirmModal>
    </EuiOverlayMask>
  );
};
