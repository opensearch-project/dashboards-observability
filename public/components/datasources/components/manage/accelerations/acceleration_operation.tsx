/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { useEffect, useState } from 'react';
import { CachedAcceleration } from '../../../../../../common/types/data_connections';
import { useToast } from '../../../../common/toast';
import { useDirectQuery } from '../../../../../framework/datasources/direct_query_hook';
import { DirectQueryLoadingStatus } from '../../../../../../common/types/explorer';
import {
  AccelerationActionType,
  generateAccelerationOperationQuery,
  getAccelerationName,
} from './utils/acceleration_utils';

export const useAccelerationOperation = (dataSource: string) => {
  const { startLoading, stopLoading, loadStatus } = useDirectQuery();
  const { setToast } = useToast();
  const [isOperating, setIsOperating] = useState(false);
  const [operationSuccess, setOperationSuccess] = useState(false);
  const [accelerationToOperate, setAccelerationToOperate] = useState<CachedAcceleration | null>(
    null
  );
  const [operationType, setOperationType] = useState<AccelerationActionType | null>(null);
  const [currentStatus, setCurrentStatus] = useState<DirectQueryLoadingStatus | null>(null);

  useEffect(() => {
    if (!accelerationToOperate || !operationType || loadStatus === currentStatus) return;

    const displayAccelerationName = getAccelerationName(accelerationToOperate);

    let operationInProgressMessage = '';
    let operationSuccessMessage = '';
    let operationFailureMessage = '';

    switch (operationType) {
      case 'delete':
        operationInProgressMessage = `Deleting acceleration: ${displayAccelerationName}`;
        operationSuccessMessage = `Successfully deleted acceleration: ${displayAccelerationName}`;
        operationFailureMessage = `Failed to delete acceleration: ${displayAccelerationName}`;
        break;
      case 'vacuum':
        operationInProgressMessage = `Vacuuming acceleration: ${displayAccelerationName}`;
        operationSuccessMessage = `Successfully vacuumed acceleration: ${displayAccelerationName}`;
        operationFailureMessage = `Failed to vacuum acceleration: ${displayAccelerationName}`;
        break;
      case 'sync':
        operationInProgressMessage = `Syncing acceleration: ${displayAccelerationName}`;
        break;
    }

    if (loadStatus === DirectQueryLoadingStatus.SCHEDULED && operationType !== 'sync') {
      setIsOperating(true);
      setToast(operationInProgressMessage, 'success');
    } else if (loadStatus === DirectQueryLoadingStatus.SUCCESS && operationType !== 'sync') {
      setIsOperating(false);
      setAccelerationToOperate(null);
      setOperationSuccess(true);
      setToast(operationSuccessMessage, 'success');
    } else if (loadStatus === DirectQueryLoadingStatus.FAILED && operationType !== 'sync') {
      setIsOperating(false);
      setOperationSuccess(false);
      setToast(operationFailureMessage, 'danger');
    } else if (operationType === 'sync' && loadStatus === DirectQueryLoadingStatus.SCHEDULED) {
      setToast(operationInProgressMessage, 'success');
      stopLoading();
    }

    setCurrentStatus(loadStatus);
  }, [loadStatus, setToast, accelerationToOperate, operationType, currentStatus]);

  useEffect(() => {
    return () => {
      stopLoading();
    };
  }, []);

  const performOperation = (
    acceleration: CachedAcceleration,
    operation: AccelerationActionType
  ) => {
    setOperationSuccess(false);
    setOperationType(operation);
    const operationQuery = generateAccelerationOperationQuery(acceleration, dataSource, operation);

    const requestPayload = {
      lang: 'sql',
      query: operationQuery,
      datasource: dataSource,
    };

    setIsOperating(true);
    setAccelerationToOperate(acceleration);
    startLoading(requestPayload);
  };

  return { performOperation, isOperating, operationSuccess };
};
