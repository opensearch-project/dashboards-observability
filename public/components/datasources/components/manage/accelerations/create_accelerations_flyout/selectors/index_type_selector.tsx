/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiFormRow, EuiLink, EuiSpacer, EuiSuperSelect, EuiText } from '@elastic/eui';
import React, { Fragment, useEffect, useState } from 'react';
import {
  ACCELERATION_DEFUALT_SKIPPING_INDEX_NAME,
  ACC_INDEX_TYPE_DOCUMENTATION_URL,
} from '../../../../../../../../common/constants/data_sources';
import {
  AccelerationIndexType,
  CreateAccelerationForm,
} from '../../../../../../../../common/types/data_connections';

interface IndexTypeSelectorProps {
  accelerationFormData: CreateAccelerationForm;
  setAccelerationFormData: React.Dispatch<React.SetStateAction<CreateAccelerationForm>>;
  initiateColumnLoad: (dataSource: string, database: string, dataTable: string) => void;
  loading: boolean;
}

export const IndexTypeSelector = ({
  accelerationFormData,
  setAccelerationFormData,
  initiateColumnLoad,
  loading,
}: IndexTypeSelectorProps) => {
  const [value, setValue] = useState('skipping');

  useEffect(() => {
    initiateColumnLoad(
      accelerationFormData.dataSource,
      accelerationFormData.database,
      accelerationFormData.dataTable
    );
  }, [accelerationFormData.dataTable]);

  const onChangeSupeSelect = (indexType: string) => {
    setAccelerationFormData({
      ...accelerationFormData,
      accelerationIndexType: indexType as AccelerationIndexType,
      accelerationIndexName:
        indexType === 'skipping' ? ACCELERATION_DEFUALT_SKIPPING_INDEX_NAME : '',
    });
    setValue(indexType);
  };

  const superSelectOptions = [
    {
      value: 'skipping',
      inputDisplay: 'Skipping index',
      dropdownDisplay: (
        <Fragment>
          <strong>Skipping index</strong>
          <EuiText size="s" color="subdued">
            <p className="EuiTextColor--subdued">
              Accelerate direct queries by storing table meta-data in OpenSearch.
            </p>
          </EuiText>
        </Fragment>
      ),
    },
    {
      value: 'covering',
      inputDisplay: 'Covering index',
      dropdownDisplay: (
        <Fragment>
          <strong>Covering index</strong>
          <EuiText size="s" color="subdued">
            <p className="EuiTextColor--subdued">
              Accelerate queries from subset of table data stored in OpenSearch.
            </p>
          </EuiText>
        </Fragment>
      ),
    },
    {
      value: 'materialized',
      inputDisplay: 'Materialized view',
      dropdownDisplay: (
        <Fragment>
          <strong>Materialized view</strong>
          <EuiText size="s" color="subdued">
            <p className="EuiTextColor--subdued">
              Accelerate queries and visualizations from aggregated table results stored in
              OpenSearch.
            </p>
          </EuiText>
        </Fragment>
      ),
    },
  ];

  return (
    <>
      <EuiText data-test-subj="acceleration-setting-header">
        <h3>Acceleration setting</h3>
      </EuiText>
      <EuiSpacer size="s" />
      <EuiFormRow
        label="Index type"
        helpText="Select the type of index you want to create. Each index type has benefits and costs."
        labelAppend={
          <EuiText size="xs">
            <EuiLink href={ACC_INDEX_TYPE_DOCUMENTATION_URL} target="_blank">
              Help
            </EuiLink>
          </EuiText>
        }
      >
        <EuiSuperSelect
          options={superSelectOptions}
          valueOfSelected={value}
          onChange={onChangeSupeSelect}
          itemLayoutAlign="top"
          hasDividers
          isLoading={loading}
        />
      </EuiFormRow>
    </>
  );
};
