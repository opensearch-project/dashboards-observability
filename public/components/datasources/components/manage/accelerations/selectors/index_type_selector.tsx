/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiComboBox, EuiComboBoxOptionOption, EuiFormRow, EuiLink, EuiText } from '@elastic/eui';
import React, { useState } from 'react';
import {
  ACCELERATION_DEFUALT_SKIPPING_INDEX_NAME,
  ACCELERATION_INDEX_TYPES,
  ACC_INDEX_TYPE_DOCUMENTATION_URL,
} from '../../../../../../../common/constants/data_sources';
import {
  AccelerationIndexType,
  CreateAccelerationForm,
} from '../../../../../../../common/types/data_connections';

interface IndexTypeSelectorProps {
  accelerationFormData: CreateAccelerationForm;
  setAccelerationFormData: React.Dispatch<React.SetStateAction<CreateAccelerationForm>>;
}

export const IndexTypeSelector = ({
  accelerationFormData,
  setAccelerationFormData,
}: IndexTypeSelectorProps) => {
  const [selectedIndexType, setSelectedIndexType] = useState<
    Array<EuiComboBoxOptionOption<string>>
  >([ACCELERATION_INDEX_TYPES[0]]);
  const [loading, _setLoading] = useState(false);

  // useEffect(() => {
  // TODO: Load table schema from cache
  //   if (accelerationFormData.dataTable !== '') {
  //     setLoading(true);
  //     const idPrefix = htmlIdGenerator()();
  //     const query = {
  //       lang: 'sql',
  //       query: `DESC \`${accelerationFormData.dataSource}\`.\`${accelerationFormData.database}\`.\`${accelerationFormData.dataTable}\``,
  //       datasource: accelerationFormData.dataSource,
  //     };

  //     executeAsyncQuery(
  //       accelerationFormData.dataSource,
  //       query,
  //       (response: AsyncApiResponse) => {
  //         const status = response.data.resp.status.toLowerCase();
  //         if (status === AsyncQueryStatus.Success) {
  //           const dataTableFields: DataTableFieldsType[] = response.data.resp.datarows
  //             .filter((row) => !row[0].startsWith('#'))
  //             .map((row, index) => ({
  //               id: `${idPrefix}${index + 1}`,
  //               fieldName: row[0],
  //               dataType: row[1],
  //             }));
  //           setAccelerationFormData({
  //             ...accelerationFormData,
  //             dataTableFields,
  //           });
  //           setLoading(false);
  //         }
  //         if (status === AsyncQueryStatus.Failed || status === AsyncQueryStatus.Cancelled) {
  //           setLoading(false);
  //         }
  //       },
  //       () => setLoading(false)
  //     );
  //   }
  // }, [accelerationFormData.dataTable]);

  const onChangeIndexType = (indexTypeOption: Array<EuiComboBoxOptionOption<string>>) => {
    const indexType = indexTypeOption[0].value as AccelerationIndexType;
    setAccelerationFormData({
      ...accelerationFormData,
      accelerationIndexType: indexType,
      accelerationIndexName:
        indexType === 'skipping' ? ACCELERATION_DEFUALT_SKIPPING_INDEX_NAME : '',
    });
    setSelectedIndexType(indexTypeOption);
  };
  return (
    <>
      <EuiFormRow
        label="Index type"
        helpText="Select the type of index you want to create. Each index type has benefits and costs."
        labelAppend={
          <EuiText size="xs">
            <EuiLink href={ACC_INDEX_TYPE_DOCUMENTATION_URL} target="_blank">
              Help
            </EuiLink>
          </EuiText>
        }
      >
        <EuiComboBox
          placeholder="Select an index type"
          singleSelection={{ asPlainText: true }}
          options={ACCELERATION_INDEX_TYPES}
          selectedOptions={selectedIndexType}
          onChange={onChangeIndexType}
          isInvalid={selectedIndexType.length === 0}
          isClearable={false}
          isLoading={loading}
        />
      </EuiFormRow>
    </>
  );
};
