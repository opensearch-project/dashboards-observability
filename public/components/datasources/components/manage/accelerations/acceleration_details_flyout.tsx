/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButtonEmpty,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFlyoutBody,
  EuiFlyoutHeader,
  EuiIcon,
  EuiSpacer,
  EuiTab,
  EuiTabs,
  EuiText,
} from '@elastic/eui';
import React, { useEffect, useState } from 'react';
import { OpenSearchDashboardsResponse } from '../../../../../../../../src/core/server/http/router';
import { CachedAcceleration } from '../../../../../../common/types/data_connections';
import { coreRefs } from '../../../../../framework/core_refs';
import { AccelerationActionOverlay } from './acceleration_action_overlay';
import { useAccelerationOperation } from './acceleration_operation';
import { AccelerationDetailsTab } from './flyout_modules/acceleration_details_tab';
import { AccelerationSchemaTab } from './flyout_modules/accelerations_schema_tab';
import {
  AccelerationActionType,
  getAccelerationName,
  onDiscoverIconClick,
} from './utils/acceleration_utils';

export interface AccelerationDetailsFlyoutProps {
  acceleration: CachedAcceleration;
  dataSourceName: string;
  resetFlyout: () => void;
  handleRefresh?: () => void;
  dataSourceMDSId?: string;
}

const getMappings = (index: string): Promise<OpenSearchDashboardsResponse> | undefined => {
  return coreRefs.dslService?.fetchFields(index);
};

const getSettings = (index: string): Promise<OpenSearchDashboardsResponse> | undefined => {
  return coreRefs.dslService?.fetchSettings(index);
};

const getIndexInfo = (index: string): Promise<OpenSearchDashboardsResponse> | undefined => {
  return coreRefs.dslService?.fetchIndices(index);
};

const handleDetailsFetchingPromise = (
  promise: Promise<OpenSearchDashboardsResponse> | undefined,
  action: string
) => {
  return promise!
    .then((data) => ({ status: 'fulfilled', action, data }))
    .catch((error) => ({ status: 'rejected', action, error }));
};

export const AccelerationDetailsFlyout = (props: AccelerationDetailsFlyoutProps) => {
  const { dataSourceName, acceleration, resetFlyout, handleRefresh } = props;
  const { flintIndexName } = acceleration;
  const [selectedTab, setSelectedTab] = useState('details');
  const tabsMap: { [key: string]: any } = {
    details: AccelerationDetailsTab,
    schema: AccelerationSchemaTab,
  };
  const [operationType, setOperationType] = useState<AccelerationActionType | null>(null);
  const [showConfirmationOverlay, setShowConfirmationOverlay] = useState(false);

  const { performOperation, operationSuccess } = useAccelerationOperation(props.dataSourceName);

  const displayedIndex = getAccelerationName(acceleration);

  const onConfirmOperation = () => {
    if (operationType && props.acceleration) {
      performOperation(props.acceleration, operationType);
      setShowConfirmationOverlay(false);
    }
  };

  const onSyncIconClickHandler = () => {
    setOperationType('sync');
    setShowConfirmationOverlay(true);
  };

  const onDeleteIconClickHandler = () => {
    setOperationType('delete');
    setShowConfirmationOverlay(true);
  };

  const onVacuumIconClickHandler = () => {
    setOperationType('vacuum');
    setShowConfirmationOverlay(true);
  };

  const [settings, setSettings] = useState<object>();
  const [mappings, setMappings] = useState();
  const [indexInfo, setIndexInfo] = useState();

  const updateMapping = (result) => {
    setMappings(result);
  };

  const updateSetting = (result, slectedIndex: string) => {
    setSettings(result.data[slectedIndex]);
  };

  const updateIndexInfo = (result) => {
    setIndexInfo(result);
  };

  const getAccDetail = (selectedIndex: string) => {
    Promise.all([
      handleDetailsFetchingPromise(getMappings(selectedIndex), 'getMappings'),
      handleDetailsFetchingPromise(getSettings(selectedIndex), 'getSettings'),
      handleDetailsFetchingPromise(getIndexInfo(selectedIndex), 'getIndexInfo'),
    ])
      .then((results) => {
        updateMapping(results[0]);
        updateSetting(results[1], selectedIndex);
        updateIndexInfo(results[2]);
      })
      .catch((errors: Error[]) => {
        errors.forEach((error, errorIndex) => {
          console.error(`Error in async call ${errorIndex + 1}:`, error);
        });
      });
  };

  useEffect(() => {
    if (operationSuccess) {
      resetFlyout();
      handleRefresh?.();
      setOperationType(null);
      setShowConfirmationOverlay(false);
    }
  }, [operationSuccess, resetFlyout, handleRefresh]);

  useEffect(() => {
    if (flintIndexName !== undefined && flintIndexName.trim().length > 0) {
      getAccDetail(flintIndexName);
    }
  }, [flintIndexName]);

  const DiscoverIcon = () => {
    return (
      <EuiButtonEmpty
        onClick={() => {
          onDiscoverIconClick(acceleration, dataSourceName);
          resetFlyout();
        }}
      >
        <EuiIcon type={'discoverApp'} size="m" />
      </EuiButtonEmpty>
    );
  };

  const SyncIcon = ({ autoRefresh, status }: { autoRefresh: boolean; status: string }) => {
    if (autoRefresh || status !== 'active') {
      return null;
    }
    return (
      <EuiButtonEmpty onClick={onSyncIconClickHandler}>
        <EuiIcon type="inputOutput" size="m" />
      </EuiButtonEmpty>
    );
  };

  const DeleteIcon = () => {
    return (
      <EuiButtonEmpty onClick={onDeleteIconClickHandler}>
        <EuiIcon type="trash" size="m" />
      </EuiButtonEmpty>
    );
  };

  const VacuumIcon = () => {
    return (
      <EuiButtonEmpty onClick={onVacuumIconClickHandler}>
        <EuiIcon type="broom" size="m" />
      </EuiButtonEmpty>
    );
  };

  const accelerationDetailsTabs = [
    {
      id: 'details',
      name: 'Details',
      disabled: false,
    },
    {
      id: 'schema',
      name: 'Schema',
      disabled: false,
    },
  ];

  const renderTabs = () => {
    return accelerationDetailsTabs.map((tab, tabIndex) => {
      return (
        <EuiTab
          onClick={() => setSelectedTab(tab.id)}
          isSelected={tab.id === selectedTab}
          disabled={tab.disabled}
          key={tabIndex}
        >
          {tab.name}
        </EuiTab>
      );
    });
  };

  const renderTabContent = (tab: string) => {
    let propsForTab;

    switch (tab) {
      case 'details':
        propsForTab = { acceleration, settings, mappings, indexInfo, dataSourceName, resetFlyout };
        break;
      case 'schema':
        propsForTab = { mappings, indexInfo };
        break;
      default:
        return null;
    }

    const TabToDisplay = tabsMap[tab];

    return <TabToDisplay {...propsForTab} />;
  };

  return (
    <>
      <EuiFlyoutHeader hasBorder>
        <EuiFlexGroup direction="row" alignItems="center" gutterSize="m">
          <EuiFlexItem>
            <EuiText>
              <h2 className="panel-title">{displayedIndex}</h2>
            </EuiText>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <SyncIcon autoRefresh={acceleration.autoRefresh} status={acceleration.status} />
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <DiscoverIcon />
          </EuiFlexItem>
          {acceleration.status !== 'deleted' ? (
            <EuiFlexItem grow={false}>
              <DeleteIcon />
            </EuiFlexItem>
          ) : (
            <EuiFlexItem grow={false}>
              <VacuumIcon />
            </EuiFlexItem>
          )}
        </EuiFlexGroup>
        <EuiSpacer size="m" />
        <EuiTabs style={{ marginBottom: '-25px' }}>{renderTabs()}</EuiTabs>
      </EuiFlyoutHeader>
      <EuiFlyoutBody>{renderTabContent(selectedTab)}</EuiFlyoutBody>
      {showConfirmationOverlay && operationType && (
        <AccelerationActionOverlay
          isVisible={showConfirmationOverlay}
          actionType={operationType as AccelerationActionType}
          acceleration={acceleration}
          dataSourceName={dataSourceName}
          onCancel={() => setShowConfirmationOverlay(false)}
          onConfirm={onConfirmOperation}
        />
      )}
    </>
  );
};
