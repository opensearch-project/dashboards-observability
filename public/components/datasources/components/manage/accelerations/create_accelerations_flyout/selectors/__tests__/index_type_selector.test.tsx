/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { waitFor } from '@testing-library/dom';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import React from 'react';
import { CreateAccelerationForm } from '../../../../../../../../../common/types/data_connections';
import { createAccelerationEmptyDataMock } from '../../../../../../../../../test/accelerations';
import { IndexTypeSelector } from '../index_type_selector';

describe('Index type selector components', () => {
  configure({ adapter: new Adapter() });

  it('renders type selector with default options', async () => {
    const accelerationFormData = createAccelerationEmptyDataMock;
    const setAccelerationFormData = jest.fn();
    const wrapper = mount(
      <IndexTypeSelector
        accelerationFormData={accelerationFormData}
        dataSourceType="S3GLUE"
        setAccelerationFormData={setAccelerationFormData}
        initiateColumnLoad={jest.fn()}
      />
    );
    wrapper.update();
    await waitFor(() => {
      expect(
        toJson(wrapper, {
          noKey: false,
          mode: 'deep',
        })
      ).toMatchSnapshot();
    });
  });

  it('renders type selector with different options', async () => {
    const accelerationFormData: CreateAccelerationForm = {
      ...createAccelerationEmptyDataMock,
      accelerationIndexType: 'covering',
    };
    const setAccelerationFormData = jest.fn();
    const wrapper = mount(
      <IndexTypeSelector
        accelerationFormData={accelerationFormData}
        dataSourceType="S3GLUE"
        setAccelerationFormData={setAccelerationFormData}
        initiateColumnLoad={jest.fn()}
      />
    );
    wrapper.update();
    await waitFor(() => {
      expect(
        toJson(wrapper, {
          noKey: false,
          mode: 'deep',
        })
      ).toMatchSnapshot();
    });
  });
});
