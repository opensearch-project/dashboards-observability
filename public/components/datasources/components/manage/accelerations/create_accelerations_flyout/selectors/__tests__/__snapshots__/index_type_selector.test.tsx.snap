// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index type selector components renders type selector with default options 1`] = `
Array [
  <div
    className="euiText euiText--medium"
    data-test-subj="acceleration-setting-header"
  >
    <h3>
      Acceleration setting
    </h3>
  </div>,
  <div
    className="euiSpacer euiSpacer--s"
  />,
  <div
    className="euiFormRow"
    id="random_html_id-row"
  >
    <div
      className="euiFormRow__labelWrapper"
    >
      <label
        className="euiFormLabel euiFormRow__label"
        htmlFor="random_html_id"
      >
        Index type
      </label>
       
      <div
        className="euiText euiText--extraSmall"
      >
        <a
          className="euiLink euiLink--primary"
          href="https://github.com/opensearch-project/opensearch-spark/blob/main/docs/index.md"
          rel="noopener noreferrer"
          target="_blank"
        >
          Help
          <svg
            aria-hidden={true}
            aria-label="External link"
            className="euiIcon euiIcon--small euiIcon-isLoading euiLink__externalIcon"
            focusable="false"
            height={16}
            role="img"
            style={null}
            viewBox="0 0 16 16"
            width={16}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
            />
          </svg>
          <span
            className="euiScreenReaderOnly"
          >
            (opens in a new tab or window)
          </span>
        </a>
      </div>
    </div>
    <div
      className="euiFormRow__fieldWrapper"
    >
      <div
        className="euiPopover euiPopover--anchorDownLeft euiPopover--displayBlock euiInputPopover euiSuperSelect"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="euiPopover__anchor"
        >
          <div>
            Array [
              <input
                id="random_html_id"
                type="hidden"
                value="skipping"
              />,
              <div
                className="euiFormControlLayout"
              >
                <div
                  className="euiFormControlLayout__childrenWrapper"
                >
                  <span
                    className="euiScreenReaderOnly"
                    id="random_html_id"
                  >
                    Select an option: Skipping index, is selected
                  </span>
                  <button
                    aria-describedby="random_html_id-help-0"
                    aria-haspopup="true"
                    aria-labelledby="random_html_id random_html_id"
                    className="euiSuperSelectControl"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    type="button"
                  >
                    Skipping index
                  </button>
                  <div
                    className="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                  >
                    <span
                      className="euiFormControlLayoutCustomIcon"
                    >
                      <svg
                        aria-hidden={true}
                        className="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
                        focusable="false"
                        height={16}
                        role="img"
                        style={null}
                        viewBox="0 0 16 16"
                        width={16}
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                        />
                      </svg>
                    </span>
                  </div>
                </div>
              </div>,
            ]
          </div>
        </div>
      </div>
      <div
        className="euiFormHelpText euiFormRow__text"
        id="random_html_id-help-0"
      >
        Select the type of index you want to create. Each index type has benefits and costs.
      </div>
    </div>
  </div>,
]
`;

exports[`Index type selector components renders type selector with different options 1`] = `
Array [
  <div
    className="euiText euiText--medium"
    data-test-subj="acceleration-setting-header"
  >
    <h3>
      Acceleration setting
    </h3>
  </div>,
  <div
    className="euiSpacer euiSpacer--s"
  />,
  <div
    className="euiFormRow"
    id="random_html_id-row"
  >
    <div
      className="euiFormRow__labelWrapper"
    >
      <label
        className="euiFormLabel euiFormRow__label"
        htmlFor="random_html_id"
      >
        Index type
      </label>
       
      <div
        className="euiText euiText--extraSmall"
      >
        <a
          className="euiLink euiLink--primary"
          href="https://github.com/opensearch-project/opensearch-spark/blob/main/docs/index.md"
          rel="noopener noreferrer"
          target="_blank"
        >
          Help
          <svg
            aria-label="External link"
            className="euiIcon euiIcon--small euiLink__externalIcon"
            focusable="false"
            height={16}
            role="img"
            style={null}
            viewBox="0 0 16 16"
            width={16}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M13 8.5a.5.5 0 1 1 1 0V12a2 2 0 0 1-2 2H4a2 2 0 0 1-2-2V4a2 2 0 0 1 2-2h3.5a.5.5 0 0 1 0 1H4a1 1 0 0 0-1 1v8a1 1 0 0 0 1 1h8a1 1 0 0 0 1-1V8.5Zm-5.12.339a.5.5 0 1 1-.706-.707L13.305 2H10.5a.5.5 0 1 1 0-1H14a1 1 0 0 1 1 1v3.5a.5.5 0 1 1-1 0V2.72L7.88 8.838Z"
            />
          </svg>
          <span
            className="euiScreenReaderOnly"
          >
            (opens in a new tab or window)
          </span>
        </a>
      </div>
    </div>
    <div
      className="euiFormRow__fieldWrapper"
    >
      <div
        className="euiPopover euiPopover--anchorDownLeft euiPopover--displayBlock euiInputPopover euiSuperSelect"
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
      >
        <div
          className="euiPopover__anchor"
        >
          <div>
            Array [
              <input
                id="random_html_id"
                type="hidden"
                value="skipping"
              />,
              <div
                className="euiFormControlLayout"
              >
                <div
                  className="euiFormControlLayout__childrenWrapper"
                >
                  <span
                    className="euiScreenReaderOnly"
                    id="random_html_id"
                  >
                    Select an option: Skipping index, is selected
                  </span>
                  <button
                    aria-describedby="random_html_id-help-0"
                    aria-haspopup="true"
                    aria-labelledby="random_html_id random_html_id"
                    className="euiSuperSelectControl"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    type="button"
                  >
                    Skipping index
                  </button>
                  <div
                    className="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                  >
                    <span
                      className="euiFormControlLayoutCustomIcon"
                    >
                      <svg
                        aria-hidden={true}
                        className="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
                        focusable="false"
                        height={16}
                        role="img"
                        style={null}
                        viewBox="0 0 16 16"
                        width={16}
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M13.069 5.157 8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0Z"
                          fillRule="non-zero"
                        />
                      </svg>
                    </span>
                  </div>
                </div>
              </div>,
            ]
          </div>
        </div>
      </div>
      <div
        className="euiFormHelpText euiFormRow__text"
        id="random_html_id-help-0"
      >
        Select the type of index you want to create. Each index type has benefits and costs.
      </div>
    </div>
  </div>,
]
`;
